license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/CopyrightRangeProvider.java

```diff
@@ -76,50 +78,60 @@ public void close() {
    * {@link RuntimeException} is thrown.
    */
   @Override
-  public Map<String, String> adjustProperties(AbstractLicenseMojo mojo,
-                                              Map<String, String> properties, Document document) {
+  public Map<String, Supplier<String>> adjustLazyProperties(AbstractLicenseMojo mojo, Map<String, String> properties, Document document) {
     String inceptionYear = properties.get(INCEPTION_YEAR_KEY);
     if (inceptionYear == null) {
-      throw new RuntimeException("'" + INCEPTION_YEAR_KEY + "' must have a value for file "
-          + document.getFile().getAbsolutePath());
+      throw new RuntimeException("'" + INCEPTION_YEAR_KEY + "' must have a value for file " + document.getFile().getAbsolutePath());
     }
     final int inceptionYearInt;
     try {
       inceptionYearInt = Integer.parseInt(inceptionYear);
     } catch (NumberFormatException e1) {
       throw new RuntimeException(
-          "'" + INCEPTION_YEAR_KEY + "' must be an integer ; found = " + inceptionYear + " file: "
-              + document.getFile().getAbsolutePath());
+          "'" + INCEPTION_YEAR_KEY + "' must be an integer ; found = " + inceptionYear + " file: " + document.getFile().getAbsolutePath());
     }
-    try {
-      Map<String, String> result = new HashMap<>(4);
 
-      int copyrightEnd = gitLookup.getYearOfLastChange(document.getFile());
-      result.put(COPYRIGHT_LAST_YEAR_KEY, Integer.toString(copyrightEnd));
-      final String copyrightYears;
-      if (inceptionYearInt >= copyrightEnd) {
-        copyrightYears = inceptionYear;
-      } else {
-        copyrightYears = inceptionYear + "-" + copyrightEnd;
-      }
-      result.put(COPYRIGHT_YEARS_KEY, copyrightYears);
+    final var memo = new HashMap<String, Integer>();
+    final Function<String, Integer> memoGet = (k) -> {
+      return memo.computeIfAbsent(k, (key) -> {
+        try {
+          if (COPYRIGHT_LAST_YEAR_KEY.equals(key)) {
+            return gitLookup.getYearOfLastChange(document.getFile());
+          } else if (COPYRIGHT_CREATION_YEAR_KEY.equals(key)) {
+            return gitLookup.getYearOfCreation(document.getFile());
+          }
+          return null;
+        } catch (IOException | GitAPIException e) {
+          throw new RuntimeException(
+              "CopyrightRangeProvider error on file: " + document.getFile().getAbsolutePath() + ": " + e.getMessage(), e);
+        }
+      });
+    };
```

Comment:
Even if I understand the memoizing effect of it, I think it can be written in a better form for code readability and future maintainance by introduce a static method:

```
    public static <T, R> Function<T, R> memoize(Function<T, R> function) {
        Map<T, R> cache = new HashMap<>();
        return input -> cache.computeIfAbsent(input, function);
    }
```

and then crating only 2 functions for the 2 cases.

Reply:
Unfortunately, Java's missing approach to checked exceptions for Lambdas prevents this slick solution.
I went with an atomic ref and a helper function to avoid code duplication.

Reply:
There is a way to hack around that using templates for exceptions but this is a hacky way lol ;-) but since not a public API would be acceptable. The atomic ref is fine also ;-)